src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Helpers.cs

```diff
@@ -89,34 +89,82 @@ public static X509Certificate2 TryFindMatchingCertificate(this X509Certificate2C
                 case SubjectIdentifierType.IssuerAndSerialNumber:
                     {
                         X509IssuerSerial issuerSerial = (X509IssuerSerial)(recipientIdentifier.Value);
-
-                        using (ClonedCertificates matches1 = certs.FindCertificates(X509FindType.FindBySerialNumber, issuerSerial.SerialNumber))
+                        string serialNumber = issuerSerial.SerialNumber;
+                        string issuer = issuerSerial.IssuerName;
+                        foreach (X509Certificate2 candidate in certs)
                         {
-                            using (ClonedCertificates matches2 = matches1.FindCertificates(X509FindType.FindByIssuerDistinguishedName, issuerSerial.IssuerName))
-                            {
-                                // Desktop compat: We do not complain about multiple matches. Just take the first one and ignore the rest.
-                                return matches2.Any ? matches2.ClaimResult() : null;
-                            }
+                            if (candidate.SerialNumber == serialNumber && candidate.Issuer == issuer)
+                                return candidate;
                         }
                     }
+                    break;
 
                 case SubjectIdentifierType.SubjectKeyIdentifier:
                     {
-                        string ski = (string)(recipientIdentifier.Value);
-                        using (ClonedCertificates matches = certs.FindCertificates(X509FindType.FindBySubjectKeyIdentifier, ski))
+                        string skiString = (string)(recipientIdentifier.Value);
+                        byte[] ski = skiString.ToSki();
+                        foreach (X509Certificate2 cert in certs)
                         {
-                            // Desktop compat: We do not complain about multiple matches. Just take the first one and ignore the rest.
-                            return matches.Any ? matches.ClaimResult() : null;
+                            byte[] candidateSki = PkcsPal.Instance.GetSubjectKeyIdentifier(cert);
```

Comment:
Hmm… looking at X509SubjectKeyIdentifierExtension, it only wants to give me back the SKI in stringized form. I’d rather compare byte arrays, especially since both PAL’s need the byte array form anyway for the encoding part – and the CERT_KEY_IDENTIFIER_PROP_ID already gives me what I want. So for Windows, at least, this is still the most convenient factoring.


From: Jeremy Barton [mailto:notifications@github.com]
Sent: Thursday, May 5, 2016 2:25 PM
To: dotnet/corefx <corefx@noreply.github.com>
Cc: Atsushi Kanamori <atsushik@microsoft.com>; Author <author@noreply.github.com>
Subject: Re: [dotnet/corefx] Fix EnvelopeCms Decrypt() failure on collection-imported PFXs. (#8294)


In src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Helpers.cs<https://github.com/dotnet/corefx/pull/8294#discussion_r62257698>:

>                          {

> -                            // Desktop compat: We do not complain about multiple matches. Just take the first one and ignore the rest.

> -                            return matches.Any ? matches.ClaimResult() : null;

> +                            byte[] candidateSki = PkcsPal.Instance.GetSubjectKeyIdentifier(cert);

Why send this to the PAL instead of reading the cert extensions?

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/dotnet/corefx/pull/8294/files/2772d787f16f24ef6b1f1c5de097aae9a565712f#r62257698>